torch/csrc/generic/StorageSharing.cpp

```diff
@@ -101,7 +101,7 @@ static PyObject * THPStorage_(shareFilename)(THPStorage *self)
     ctx = (libshm_context*)allocator_obj->allocatorContext;
   } else {
     // TODO: retry on collision
-    // TODO: free GIL - but remember to reacquire it when an exception is thrown
+    AutoNoGIL no_gil;
```

Comment:
It wasn't thread safe even with the GIL: any concurrent read operations on this storage would cause a data race.

I'm not a huge fan of this sort of ad-hoc locking, since it gives the appearance of thread safety without the guarantees. So far we've said that concurrent read operations are thread-safe, but mutating ops (`_`-ending functions) are not safe concurrently with each other or with read ops.

I also would prefer to avoid a single global mutex around these ops because that prevents threads from copying unrelated storages to shared memory in parallel.

We could narrow the AutoNoGIL to around the storage construction and copy but exclude the `swap`. That would lessen the chance of metadata corruption, although it still wouldn't be thread-safe.

Reply:
Ok, I think that if we're just saying that "it's not thread safe" then it doesn't make sense to protect just the `swap`. Let's leave it as is